app/views/compute_resources_vms/form/vmware/_base.html.erb

```
@@ -49,6 +50,14 @@ end %>
                { :disabled => images.empty?, :label => _('Image'), :label_size => "col-md-2" } %>
 </div>
 
+<!-- # BIOS firmware doesn't support virtual TPM. -->
+<div id='efi_features' <%= 'class=hide' if !new_vm || f.object.firmware == 'bios' %>>
```

Do I get it right you can add and remove the virtual TPM module from an existing VM?

One workflow I wonder about: assume the user selected a BIOS PXE loader (which now results in a BIOS firmware) and also selects the TPM module. What happens?

---

The opposite. We hide this option if it is an existing VM or if the firmware is set to BIOS. 
However, there's an edge case where the firmware is set to Automatic, then the TPM option is available, and if you select it and set the PXE Loader to BIOS, you will get the following error:
`Failed to power up a compute VMware_with_keys (VMware) instance <HOST NAME>: GenericVmConfigFault: The virtual machine must use EFI firmware.`

---

> The opposite. We hide this option if it is an existing VM or if the firmware is set to BIOS.

`f.object.firmware` can only be set on an existing VM. If you always hide it on an existing VM, isn't it redundant? In other words:
```suggestion
<div id='efi_features' <%= 'class=hide' if !new_vm %>>
```

> However, there's an edge case where the firmware is set to Automatic, then the TPM option is available, and if you select it and set the PXE Loader to BIOS, you will get the following error:

This is what I was getting at. At first, I'd catch the situation in code before it's even sent to VMware so you can raise an error in the form. Then you can decide if you want to add some front end code to handle this. It may be useful given the two form controls are on different tabs but I wouldn't block on it when there's a proper handling post-submit.

---

> f.object.firmware can only be set on an existing VM. If you always hide it on an existing VM, isn't it redundant?

Actually, I was trying to address another edge case scenario where the user chooses the BIOS firmware. If they encounter an error when clicking submit, the firmware will be set to BIOS, and we don't want to display the TPM option.

I will handle this error on the backend as well.

---

@chris1984 I wouldn't block testing because of it. Users should know that the TPM option is not available with BIOS. I think this is an edge case (the user would have to set the firmware type to automatic, the PXE loader to BIOS, and enable the TPM option), which would be nice to cover but shouldn't block us from testing.